Utils/Dataflow/095_datasetInfoAMI/amiDatasets.py

```diff
@@ -182,9 +182,18 @@ def amiPhysValues(data):
     if not container:
         return False
     ami_client = get_ami_client()
-    res = ami_client.execute(['GetPhysicsParamsForDataset',
-                              '--logicalDatasetName=%s' % container],
-                             format='json')
+    exec_args = ['GetPhysicsParamsForDataset',
+                 '--logicalDatasetName=%s' % container]
+    if container[:4] in ['mc15', 'mc16']:
+        # Scope must be set correctly, it is 'mc15' by default.
+        if container[3] == '6':
+            exec_args.append('-scope=mc16')
```

Comment:
This looks very much like a dirty, hard-coded crutch.

* Too much magic: `[:4]`, `['mc15', 'mc16']`, `[3]`, `'6'`:
  * not clear, what does this list represent;
  * not clear, when to extend the list and when add `elif` clause;
  * not clear, why exactly 4 characters *(yes, the list contains only two 4-chars items, but no one knows it in advance)*;
  * not clear, why it's important that "the 3rd character is '6'" *(yes, it is in fact a check for "container name starting with `mc16`" -- but it is to be deducted)*;
  * very easy to break everything trying to change something.
* When AMI all of a sudden change the default scope *(by the way, haven't they yet?)*, current version will be incorrect.

Also, this code says that we only know one thing: "if the container name starts with `mc16` -- its scope is `mc16`". Maybe there is some more general rule about "what is a *scope* and how to get it from the container name", like the one we have [for the Rucio scope](https://github.com/PanDAWMS/dkb/blob/master/Utils/Dataflow/091_datasetsRucio/datasets_processing.py#L289-L307)? Will look a bit more complicated, but if we start talking about "scope" in terms of AMI -- would be great to know for sure what we're talking about and not return to this question ever again.

Reply:
Fair points.

I looked through AMI docs and various other sources of info, found no mentions of specific, AMI-related meaning of scope, in every case scope (apart from that letter from AMI team) pretty much boils down to what's implemented in our function you've linked. Furthermore, `GetPhysicsParamsForDataset` accepts "full" scopes (`mc16_13TeV`) just as fine as it does "short" ones (`mc16`).

Altogether, this means that we should probably move the function out of 091 and into a library of some sort (`Utils/Dataflow/lib/atlas.py`, for example), from which both 091 and 095 would be able to utilize it. Doing so (together with some minor improvements) would deal with most of the magic, but we'll still have to keep `['mc15', 'mc16']` in some form, because they come directly from the letter.

Thoughts? Objections?

Reply:
> I looked through AMI docs and <scope definition> pretty much boils down to what's implemented in our function...

That's a good news, thank you.

> we should probably move the function out of 091 and into a library of some sort (`Utils/Dataflow/lib/atlas.py`, for example), from which both 091 and 095 would be able to utilize it. 

If you check #371 ( *which, by the way, you have reviewed ;)* )  -- the suggested location for this functionality was in `pyDKB.atlas` module. But if you think it'd be better (*how?*), we can move it to a separate library.

> still have to keep `['mc15', 'mc16']` in some form

Of course, it is just the same situation as in the mentioned above #371 with "data formats we should ignore" (a0adae12). My point is that this list of values needs a name.

Reply:
All done. Please, take a look again.